README.md

```diff
@@ -1,11 +1,458 @@
+- `message`: the second argument to the failing check macro, or "" if
+  none was passed.
+
+If, against our advice (see [recommendation 1](#recommendations)) you
+decide not to terminate the program in response to a contract
+violation, you will omit `[[noreturn]]` and/or `noexcept`, and define
+the corresponding [preprocessor
+symbols](#contract-handler-definition).
+
+In your release builds you may wish to use a more minimal inline
+violation handler, in which case you'll define
+[`ADOBE_CONTRACT_VIOLATED_INLINE_BODY`](#contract-handler-definition).
+
+## Recommendations
+
+1. Use a contract violation handler (`adobe::contract_violated`) that
+  unconditionally terminates the program (rationale: see [About
+  Defensive Programming](#about-defensive-programming)). The
+  predefined contract violation handlers provided by this library
+  follow this recommendation.
+
+2. Start by checking whatever you can, and worry about performance
+   later. Checks are often critical for safety. [Configuration
+   options](#configuration) can be used to mitigate or eliminate costs
+   later if necessary.
+
+3. If you have to prioritize, precondition checks are the most
+   important; they are your last line of defense against undefined
+   behavior.  Postcondition checks overlap somewhat with what will be
+   checked by unit tests.  They still provide value because unit tests
+   don't cover all possible inputs and the checks may fire outside of
+   testing.
+
+   Class invariant checks can give you the most bang for your buck
+   because they can be used to eliminate the need for precondition
+   checks and verbose documentation across many functions.
+
+   ```c++
+   // Returns the day of the week corresponding to the date described
+   // by "<year>-<month>-<day>" (interpreted in ISO standard date
+   // format)
+   day_of_the_week day(int year, int month, int day) {
+     ADOBE_PRECONDITION(is_valid_date(year, month, day));
+     // implementation starts here.
+   }
+
+   // Returns the day of the week corresponding to `d`.
+   day_of_the_week day(date d) {
+     // implementation starts here.
+   }
+   ```
+
+   The second function above benefits by accepting a `date` type whose
+   invariant ensures its validity.
+
+4. The conditions in your checks should not have side-effects that
+   change program behavior, because checks are sometimes turned off by
+   configuration.
+
+2. Group all precondition checks immediately after a function's
+   opening brace, and don't allow any code to sneak in before them.
+
+3. Group all postcondition checks just before your function returns.
+   (That may mean temporarily storing a return value in a local
+   variable so it can be tested.)
+
+4. Give your `struct` or `class` a `void check_invariant() const`
+   method containing `ADOBE_INVARIANT` invocations.  Invoke it just
+   before returning from each public mutating friend or member
+   function.
```

Comment:
> and before passing `this` as an argument to an external component from a mutating operation.

Reply:
using the suggestion feature would be appreciated here too.